poseidon/src/poseidon_hash.rs

```diff
@@ -1,242 +1,71 @@
-    loop {
-        for &item in perm.squeeze() {
-            outputs.push(item);
-            if outputs.len() == num_outputs {
-                return outputs;
-            }
-        }
-        perm.permute();
-    }
+pub fn hash_n_to_m_no_pad<F: PoseidonField>(inputs: &[F], num_outputs: usize) -> Vec<F> {
+    let mut challenger = DefaultChallenger::<F, F::T>::new_poseidon_default();
+    challenger.observe_slice(inputs);
+    challenger.sample_vec(num_outputs)
 }
 
-pub fn hash_n_to_m_no_pad_iter<'a, F: PrimeField, I: Iterator<Item = &'a F>, Mds>(
-    mut input_iter: I,
+pub fn hash_n_to_m_no_pad_ext<F: PoseidonField, E: ExtensionField<BaseField = F>>(
+    inputs: &[E],
     num_outputs: usize,
-) -> Vec<F>
-where
-    Mds: MdsPermutation<F, SPONGE_WIDTH> + Default,
-{
-    let mut perm = PoseidonPermutation::<F, Mds>::new(core::iter::repeat(F::ZERO));
-
-    // Absorb all input chunks.
-    loop {
-        let chunk = input_iter.by_ref().take(SPONGE_RATE).collect::<Vec<_>>();
-        if chunk.is_empty() {
-            break;
-        }
-        // Overwrite the first r elements with the inputs. This differs from a standard sponge,
-        // where we would xor or add in the inputs. This is a well-known variant, though,
-        // sometimes called "overwrite mode".
-        perm.set_from_slice(chunk.into_iter().copied().collect::<Vec<F>>().as_slice(), 0);
-        perm.permute();
-    }
+) -> Vec<F> {
+    let mut challenger = DefaultChallenger::<F, F::T>::new_poseidon_default();
+    challenger.observe_ext_slice(inputs);
+    challenger.sample_vec(num_outputs)
+}
 
-    // Squeeze until we have the desired number of outputs
-    let mut outputs = Vec::with_capacity(num_outputs);
-    loop {
-        for &item in perm.squeeze() {
-            outputs.push(item);
-            if outputs.len() == num_outputs {
-                return outputs;
-            }
-        }
-        perm.permute();
-    }
+pub fn hash_n_to_hash_no_pad<F: PoseidonField>(inputs: &[F]) -> Digest<F> {
+    hash_n_to_m_no_pad(inputs, DIGEST_WIDTH).try_into().unwrap()
 }
 
-pub fn hash_n_to_hash_no_pad<F: PrimeField, Mds>(inputs: &[F]) -> Digest<F>
-where
-    Mds: MdsPermutation<F, SPONGE_WIDTH> + Default,
-{
-    hash_n_to_m_no_pad::<F, Mds>(inputs, DIGEST_WIDTH)
+pub fn hash_n_to_hash_no_pad_ext<F: PoseidonField, E: ExtensionField<BaseField = F>>(
+    inputs: &[E],
+) -> Digest<F> {
+    hash_n_to_m_no_pad_ext(inputs, DIGEST_WIDTH)
         .try_into()
         .unwrap()
 }
 
-pub fn compress<F: PrimeField, Mds>(x: &Digest<F>, y: &Digest<F>) -> Digest<F>
-where
-    Mds: MdsPermutation<F, SPONGE_WIDTH> + Default,
-{
-    let mut perm = PoseidonPermutation::<F, Mds>::new(core::iter::repeat(F::ZERO));
-    perm.set_from_slice(x.elements(), 0);
-    perm.set_from_slice(y.elements(), DIGEST_WIDTH);
-
-    perm.permute();
-
-    Digest(perm.squeeze()[..DIGEST_WIDTH].try_into().unwrap())
-}
-
-#[cfg(test)]
-mod tests {
```

Comment:
unittest here is to verify self-crafted poseidon impl matched with p2 poseidon. Since we all upgrade to p3 without self-crafted logic, these tests can be clean up